---
title: "2021_SAGE_II_group9_PhageDefence"
author: "Vincent Somerville"
date: "13/04/2021"
output:
  #pdf_document: default
  html_document: default
editor_options: 
  chunk_output_type: console
---

```{r setup ,echo=FALSE}
##-----------------THIS IS TO SETUP THE RMKARDOWN (WILL NOT BE PRINTED)
knitr::opts_chunk$set(message=FALSE, echo=TRUE, eval=FALSE)

```

### Recap Wally server

It has been some time since you entered the wally server. 
Here, are some refresher commands:

```{bash}
##------LOGIN to wally
ssh vsomervi@wally-front1.unil.ch
#now enter your passwrod

##------copy something to wally
scp file.txt vsomervi@wally-front1.unil.ch:/users/vsomervi/scratch/vsomervi/ # a single file
scp -r directory/ vsomervi@wally-front1.unil.ch:/users/vsomervi/scratch/vsomervi/ # a complete directory

##------navigate to SAGE II directory
cd /users/vsomervi/scratch/common_files/SAGE_II

##------look what is in there
ls 
ls -l

##------open README.txt (and close by pressing q)
less README.txt

```



## Introduction rMarkdown 

This is an rMarkdown!
rMarkdown is a coding format including both written text blocks combined with code chunks and also programming output. 
All of this is formatted by a soft latex style. 
Interestingly, you can combined not only r code but also bash, python or whatever programming language you prefer. 
To find out more see [here](https://rmarkdown.rstudio.com/) or the [cheat sheet](https://www.rstudio.com/wp-content/uploads/2015/02/rmarkdown-cheatsheet.pdf).

## Introduction git/github

Git is a version control tool. You can think of it as if you are continuous back-upping at regular intervals and you can return to certain stages of a coding project if needed. Github is the online platform that you can use to share and publish the code. 
You can sign up for [free](https://github.com/) and download and create your own content. 
I will save, update and distribute my code and presentations in the [my SAGE2 repository](https://github.com/Freevini/SAGE2_PhageDefense).
Let's try to organize the project with it. I will introduce it in more detail in a following SAGE session. 


## Reminder Wally submission


Here is the complete script that you need to run. Don't forget to change names and direcotries

```{bash echo=TRUE}
#!/bin/bash

####--------------------------------------
##SLURM options
####--------------------------------------
#SBATCH --job-name test_wally
#SBATCH --partition wally
#SBATCH --nodes 1
#SBATCH --ntasks 1
#SBATCH --cpus-per-task 2
#SBATCH --mem 2G
#SBATCH --time 03:00:00
#SBATCH --output /users/vsomervi/scratch/vsomervi/logs/%x_%j.out
#SBATCH --error /users/vsomervi/scratch/vsomervi/logs/%x_%j.err

####--------------------------------------
##preparation
##set you bash variables in order to quickly call them in the script
####--------------------------------------
username=vsomervi

genesss=$(echo "type I restriction") # gene name
genome=Lapis_ESL0263 # genome name

personal_home_directory=/users/${username}/scratch/${username}
Overall_output_directory=${personal_home_directory}/SAGE_II

####---------------------
##modules
####--------------------------------------

module load HPC/Software 
#module load UHTS/Analysis/prokka/1.13
#module load UHTS/Analysis/barrnap/0.8
module avail
module load SequenceAnalysis/HMM-Profile/hmmer/3.1b2

####--------------------------------------
##start of script
####--------------------------------------

start=$SECONDS
echo "Step: Test"

echo "=========================================================="
date +"START : %a %b %e %Y %H:%M:%S "
echo -e "Hello " ${username}

duration=$(( SECONDS - start ))
echo -e "The script ran for "${duration} "seconds"
```

## Let's do biology

Technical aspects aside let's look at the biology. 
If we want to find a specific mechanisms or pathway in our genomes what can we do?


## Look at annotations

The first step to find out if a gene is present or absent is look at the annotations. 
A useful function to scan annotations if you remember is grep. 
With grep you can scan your annotation .gff files of the Lapis_ESL0263 genome file as follows:

```{bash echo=TRUE}
grep "type I restriction" -i /users/vsomervi/scratch/common_files/SAGE_II/Genomes/Lapis_ESL0263/*gff
##-------------OR---------------------------------------------
genesss=$(echo "type I restriction") # gene name
grep "${genesss}" -i /users/vsomervi/scratch/common_files/SAGE_II/Genomes/Lapis_ESL0263/*gff
##-------------OR---------------------------------------------
genesss=$(echo "type I restriction") # gene name
genome=Lapis_ESL0263 # genome name
grep "${genesss}" -i /users/vsomervi/scratch/common_files/SAGE_II/Genomes/${genome}/*gff
```

How can we also output our results appropriately? 
In the end we want to know if the gene if present or not? (and if present is it multicopy?) 
We can do this for example with if else statements. Have a look [here](https://linuxize.com/post/bash-if-else-statement/) how it goes. 

```{bash}
genesss=$(echo "type I restriction")
genome=Lapis_ESL0263
grep "${genesss}" -i /users/vsomervi/scratch/common_files/SAGE_II/Genomes/${genome}/*gff
geneCount=$(grep -c "${genesss}" -i /users/vsomervi/scratch/common_files/SAGE_II/Genomes/${genome}/*gff) # with grep -c we count the number of gene occurrences
mkdir -p /users/vsomervi/scratch/vsomervi/SAGE_II/01_greps_gff/ # create directory 
echo -e "genome\tgene\tcount" > /users/vsomervi/scratch/vsomervi/SAGE_II/01_greps_gff/output_grep.txt # prepare output file with headers

if [ "$geneCount" = "0"  ] ##first if statement
        then
        echo "NO gene identified"
        echo -e ${genome}"\t"${genesss}"\t"${geneCount} >> /users/vsomervi/scratch/vsomervi/SAGE_II/01_greps_gff/output_grep.txt
elif [ "$geneCount" = "1"  ]  ##second if statement
        then
        echo "one gene identified"
        echo -e ${genome}"\t"${genesss}"\t"${geneCount} >> /users/vsomervi/scratch/vsomervi/SAGE_II/01_greps_gff/output_grep.txt
else  ##remaining cases
        echo "multiple genes identified"
        echo -e ${genome}"\t"${genesss}"\t"${geneCount} >> /users/vsomervi/scratch/vsomervi/SAGE_II/01_greps_gff/output_grep.txt
fi
```

This is a simple grep but how can be grep multiple names in a loop?
Here is an example how you can grep multiple gene names in a loop. 

```{bash}
#!/bin/bash

####--------------------------------------
##SLURM options
####--------------------------------------
#SBATCH --job-name test_wally
#SBATCH --partition wally
#SBATCH --nodes 1
#SBATCH --ntasks 1
#SBATCH --cpus-per-task 2
#SBATCH --mem 2G
#SBATCH --time 03:00:00
#SBATCH --output /users/vsomervi/scratch/vsomervi/logs/%x_%j.out
#SBATCH --error /users/vsomervi/scratch/vsomervi/logs/%x_%j.err

####--------------------------------------
##preparation
##set you bash variables in order to quickly call them in the script
####--------------------------------------
username=vsomervi

<<<<<<< HEAD
genes=$(echo "type I restriction") # gene name
=======
genesss=$(echo "type I restriction") # gene name
>>>>>>> 4ee193ff034d332b6443a7ac77c7a5ab3f1da6ee
genome=Lapis_ESL0263 # genome name

personal_home_directory=/users/${username}/scratch/${username}
Overall_output_directory=${personal_home_directory}/SAGE_II
##----------------------------------
##code
##----------------------------------

for genes in $(echo "lacZ lacR lacS lacA")
do
echo ${genes}
<<<<<<< HEAD
grep "${genes}" -i /users/vsomervi/scratch/common_files/SAGE_II/Genomes/${genome}/*gff #>> ${Overall_output_directory}/01_greps_gff/output_grep.txt
echo "======================"
done

=======
grep "${genes}" -i /users/vsomervi/scratch/common_files/SAGE_II/Genomes/Lapis_ESL0263/*gff
echo "======================"
done


>>>>>>> 4ee193ff034d332b6443a7ac77c7a5ab3f1da6ee
```

or if you have create a list of your genes of interest you can also cat the list:

```{bash}
for genes in $(cat /users/vsomervi/scratch/common_files/gene.list)
do
echo ${genes}
grep "${genes}" -i /users/vsomervi/scratch/common_files/SAGE_II/Genomes/Lapis_ESL0263/*gff
echo "======================"
done
```

finally, you do not only want to query a single genome but multiple genomes. therefore I suggest you loop through the genomes. 
Try it out.

#### What is next?

If we find a gene in our annotations this is usually a good sign that it is present (true positives). 
However if it is not there it could mean two things. 

![What could our results look like?](figures/type-1-2-errors.png)


Let's see if we can further distinguish between true and false negatives. 

# Hiden Markov Models (HMMs)

Some genes are very specific for a species or have not been previously annotated, therefore we need to create specific Hidden markov model (hmm) profiles of the genes of interest. 


### 1. Find the genes

In a first step you have to identify and find the genes of interest on NCBI. It is crucial that you choose the right gene. You do the following:

1. Ideally you refer to the paper and see if the name an accession number. 
2. If you take this accession number and query it on [NCBI](https://www.ncbi.nlm.nih.gov/) in **all databases** mode.
3. Look at the search and study the domains in the **Protein family models** section. 
4. Look at **Identical Protein Groups* section
5. Look at the **gene** section and identify the gene of interest. Download the nucleotide and protein fasta file and rename it with your gene Name.  

Find your way around NCBI. It has many different *help* or *about* pages. 

What you need to take from this section is one nucleotide and protein fasta file for every gene of interest. 
Remember to also fill in the [google.sheets](https://docs.google.com/spreadsheets/d/1Gp4cqbiYJ34fB7wbcqJBKsGxwgaMyPRrETnVK9PS_4M/edit#gid=0) with the information. 

### 2. Create a mutisequence alignement file

In order to create a HMM model we need to find multiple closely related forms of this gene so called gene family. From this we create a multisequence alignement (MSA) file for every gene-of-interest. 
With such a multisequence alignement file we can see conserved regions (often domains) and variable regions within a gene. We can also see all different variable sites that the genes from different species can have. 
There are many different ways how to find a gene-family and create MSA files. You will do either one of the two follow two:

#### 2.1 OMA

The best way to identify gene families of your genes-of-interest is to find them on a orthologous gene repository. 
One of these orthologous gene-family repositories is [OMA](https://omabrowser.org/oma/home/)

#### 2.2 Blast

However many of our genes are maybe not yet in the OMA browser so what we can do is the following:

1. [blastn](https://blast.ncbi.nlm.nih.gov/Blast.cgi) you genes-of-interest. Let's do it on protein level (=blastP)
2. Take the non-redudant (nr) protein database. 
3. from the blast results do an alignment (online)
4. Download and label .msa or .aln file according to the gene-of-interest name
5. **Also download a genome that contains the gene! (positive control**)

Once you have created a folder with a .msa file for every gene-of-interest you can continue to build a hmm model. 
You should move this to the wally to the appropriate location. If you cannot remember how, have a look here:

```{bash}
##------copy something to wally
scp file.txt vsomervi@wally-front1.unil.ch:/users/vsomervi/scratch/vsomervi/ # a single file
scp -r directory/ vsomervi@wally-front1.unil.ch:/users/vsomervi/scratch/vsomervi/ # a complete directory
```

### 3. Build HMM

In the next step we take the multisequencealignement file (msa) and creat the actual model. 
We therefore use a tool called [hmmer](http://hmmer.org/). Remember on wally you need to load the tools before you can use them. 

```{bash}
####--------------------------------------
##modules
####--------------------------------------

module load HPC/Software 
#module load UHTS/Analysis/prokka/1.13
#module load UHTS/Analysis/barrnap/0.8
#module avail
module load SequenceAnalysis/HMM-Profile/hmmer/3.1b2
```

Within the hmmer toolbox we will use the following two functions. See if you can understand what to do:

```{bash}
hmmbuild -h
hmmsearch -h
```

Here, you have some input for the hmmbuild part. Remember you should write a script and submit it to wally and not run it on the login node. 

```{bash}

####--------------------------------------
##hmmbuild
####--------------------------------------
hmmbuild -h
hmmbuild ${Overall_output_directory}/capV.hmm /data/projects/p539_crisprscope/defense_mechanisms/MSA_OMA/capV.fasta

```

### 4. Search HMM

<<<<<<< HEAD
First, we need to move the directory with the HMMs onto wally. 

```{bash}
scp -r /home/vincent/Desktop/Projects/2021_SAGE_II/04_HMMs/ vsomervi@wally-front1.unil.ch:/users/vsomervi/scratch/vsomervi/SAGE_II/ # a complete directory
```

Then we need to scan the the genomes for the hmms. 
We do this with a common [hmm tool](http://hmmer.org/)
```{bash}
module load HPC/Software 
module load SequenceAnalysis/HMM-Profile/hmmer/3.1b2
```

Before we can scan the genomes we however need to create a hmm index of our different Hmms with hmmpress.
Also checkout the help page
```{bash}
hmmpress -h
hmmpress /users/vsomervi/scratch/vsomervi/SAGE_II/04_HMMs/${hmmsss}.txt
```


After we have index/pressed our hmms we can scan the genomes for the genes of interest with hmmscan.
Also checkout the help page. Set your paramters accordingly. 
```{bash}
hmmscan -h
hmmscan --cpu 3 -E 10 --tblout /users/vsomervi/scratch/vsomervi/SAGE_II/05_Identification/${genomesss}/${hmmsss}.scan.tab -o /users/vsomervi/scratch/vsomervi/SAGE_II/05_Identification/${genomesss}/stdouts/${hmmsss}.scan   /users/vsomervi/scratch/vsomervi/SAGE_II/04_HMMs/${hmmsss}.txt /users/vsomervi/scratch/common_files/SAGE_II/Genomes/${genomesss}/*.faa
```

In order to make the output readable over many files look at the output files and consider the lines and columns of interest. 

```{bash}
cat /users/vsomervi/scratch/vsomervi/SAGE_II/05_Identification/${genomesss}/${hmmsss}.scan.tab
```

If you need some help. Have a look at this:

```{bash}
grep -v "^#" /users/vsomervi/scratch/vsomervi/SAGE_II/05_Identification/${genomesss}/${hmmsss}.scan.tab  |awk -v genomzz="$genomesss" -v hmmzz="$hmmsss" '{OFS="\t"}{print genomzz,hmmzz,$3}' >> /users/vsomervi/scratch/vsomervi/SAGE_II/05_Identification/all.scans
```

Also consider looping through the different HMMs/genes. e.g.

```{bash}
for hmmsss in $(ls /users/vsomervi/scratch/vsomervi/SAGE_II/04_HMMs/ | grep ".txt$" |sed 's/.txt//g')
  do
   #[...] blablabla
  done #end hmmsss loop
```

and genomes (excluding outgroup) e.g.

```{bash}
for genomesss in $(ls /users/vsomervi/scratch/common_files/SAGE_II/Genomes/| grep -v "outgroup")
  do
  for hmmsss in $(ls /users/vsomervi/scratch/vsomervi/SAGE_II/04_HMMs/ | grep ".txt$" |sed 's/.txt//g')
     do
      #[...] blablabla
     done #end hmmsss loop
  done #end genomesss loop
```

Note that in the outgroup folder there are an additional 7 genomes we want to look at. 
Therefore create a second nested loop for the outgroup genomes:

```{bash}
for genomesss in $(ls /users/vsomervi/scratch/common_files/SAGE_II/Genomes/outgroup/| grep -v "tar")
  do
  for hmmsss in $(ls /users/vsomervi/scratch/vsomervi/SAGE_II/04_HMMs/ | grep ".txt$" |sed 's/.txt//g')
     do
      #[...] blablabla
     done #end hmmsss loop
  done #end genomesss loop
```



Now put everything together and write a script for sbatch on wally. 
It looks something like this:

```{bash}
#!/bin/bash

####--------------------------------------
##SLURM options
####--------------------------------------
#SBATCH --job-name hmm_search_06
#SBATCH --partition wally
#SBATCH --nodes 1
#SBATCH --ntasks 1
#SBATCH --cpus-per-task 2
#SBATCH --mem 2G
#SBATCH --time 05:10:00
#SBATCH --output /users/vsomervi/scratch/vsomervi/logs/%x_%j.out
#SBATCH --error /users/vsomervi/scratch/vsomervi/logs/%x_%j.err

####--------------------------------------
##preparation
##set you bash variables in order to quickly call them in the script
####--------------------------------------
username=vsomervi


hmmsss=DruE
genomesss=Lapis_ESL0263
hmmsss=DruA
genomesss=Lbombicola_ESL0230
####---------------------
##modules
####--------------------------------------

module load HPC/Software 
module load SequenceAnalysis/HMM-Profile/hmmer/3.1b2

####--------------------------------------
##start of script
####--------------------------------------

start=$SECONDS
echo "Step: Test"

echo "=========================================================="
date +"START : %a %b %e %Y %H:%M:%S "
#echo -e "Hello " ${username}
rm /users/vsomervi/scratch/vsomervi/SAGE_II/05_Identification/all.scans
rm  /users/vsomervi/scratch/vsomervi/SAGE_II/05_Identification/all.scans_oneline
for genomesss in $(ls /users/vsomervi/scratch/common_files/SAGE_II/Genomes/| grep -v "outgroup")
do

rm -r /users/vsomervi/scratch/vsomervi/SAGE_II/05_Identification/${genomesss}/
mkdir -p /users/vsomervi/scratch/vsomervi/SAGE_II/05_Identification/${genomesss}/stdouts

echo "=================================================================="
echo ${genomesss}

for hmmsss in $(ls /users/vsomervi/scratch/vsomervi/SAGE_II/04_HMMs/ | grep ".txt$" |sed 's/.txt//g')
do

echo "-----------------------------------------------------------------"
echo ${hmmsss}

hmmpress /users/vsomervi/scratch/vsomervi/SAGE_II/04_HMMs/${hmmsss}.txt
hmmscan --cpu 3 -E 10 --tblout /users/vsomervi/scratch/vsomervi/SAGE_II/05_Identification/${genomesss}/${hmmsss}.scan.tab -o /users/vsomervi/scratch/vsomervi/SAGE_II/05_Identification/${genomesss}/stdouts/${hmmsss}.scan   /users/vsomervi/scratch/vsomervi/SAGE_II/04_HMMs/${hmmsss}.txt /users/vsomervi/scratch/common_files/SAGE_II/Genomes/${genomesss}/*.faa

grep -v "^#" /users/vsomervi/scratch/vsomervi/SAGE_II/05_Identification/${genomesss}/${hmmsss}.scan.tab  |awk -v genomzz="$genomesss" -v hmmzz="$hmmsss" '{OFS="\t"}{print genomzz,hmmzz,$3}' >> /users/vsomervi/scratch/vsomervi/SAGE_II/05_Identification/all.scans


#grep -v "^#" /users/vsomervi/scratch/vsomervi/SAGE_II/05_Identification/${genomesss}/${hmmsss}.scan.tab  |awk '{OFS="\t"}{print $3}'| sed -z 's/\n/,/g;s/,$/\n/' |awk -v genomzz="$genomesss" -v hmmzz="$hmmsss" '{OFS="\t"}{print genomzz,hmmzz,$1}'  >> /users/vsomervi/scratch/vsomervi/SAGE_II/05_Identification/all.scans_oneline

done # hmmsss
done #genomessss
###--------------------------
##outgroup
#for the outgroups we do it seperately
##---------------------------
for genomesss in $(ls /users/vsomervi/scratch/common_files/SAGE_II/Genomes/outgroup/| grep -v "tar")
do

rm -r /users/vsomervi/scratch/vsomervi/SAGE_II/05_Identification/${genomesss}/
mkdir -p /users/vsomervi/scratch/vsomervi/SAGE_II/05_Identification/${genomesss}/stdouts

echo "=================================================================="
echo ${genomesss}

for hmmsss in $(ls /users/vsomervi/scratch/vsomervi/SAGE_II/04_HMMs/ | grep ".txt$" |sed 's/.txt//g')
do

echo "-----------------------------------------------------------------"
echo ${hmmsss}


hmmpress /users/vsomervi/scratch/vsomervi/SAGE_II/04_HMMs/${hmmsss}.txt
hmmscan --cpu 3 -E 10 --tblout /users/vsomervi/scratch/vsomervi/SAGE_II/05_Identification/${genomesss}/${hmmsss}.scan.tab -o /users/vsomervi/scratch/vsomervi/SAGE_II/05_Identification/${genomesss}/stdouts/${hmmsss}.scan   /users/vsomervi/scratch/vsomervi/SAGE_II/04_HMMs/${hmmsss}.txt /users/vsomervi/scratch/common_files/SAGE_II/Genomes/outgroup/${genomesss}/*.faa

grep -v "^#" /users/vsomervi/scratch/vsomervi/SAGE_II/05_Identification/${genomesss}/${hmmsss}.scan.tab  |awk -v genomzz="$genomesss" -v hmmzz="$hmmsss" '{OFS="\t"}{print genomzz,hmmzz,$3}' >> /users/vsomervi/scratch/vsomervi/SAGE_II/05_Identification/all.scans


#grep -v "^#" /users/vsomervi/scratch/vsomervi/SAGE_II/05_Identification/${genomesss}/${hmmsss}.scan.tab  |awk '{OFS="\t"}{print $3}'| sed -z 's/\n/,/g;s/,$/\n/' |awk -v genomzz="$genomesss" -v hmmzz="$hmmsss" '{OFS="\t"}{print genomzz,hmmzz,$1}'  >> /users/vsomervi/scratch/vsomervi/SAGE_II/05_Identification/all.scans_oneline

done # hmmsss
done #genomessss

duration=$(( SECONDS - start ))
echo -e "The script ran for "${duration} "seconds"

##------------------------------
## genome names
##------------------------------
ls /users/vsomervi/scratch/common_files/SAGE_II/Genomes/ |grep -v "outgroup" > /users/vsomervi/scratch/vsomervi/SAGE_II/05_Identification/all.genomes
ls /users/vsomervi/scratch/common_files/SAGE_II/Genomes/outgroup/| grep -v "tar" >> /users/vsomervi/scratch/vsomervi/SAGE_II/05_Identification/all.genomes

```


The same with hmm search

```{bash}
#!/bin/bash

####--------------------------------------
##SLURM options
####--------------------------------------
#SBATCH --job-name hmm_search_06
#SBATCH --partition wally
#SBATCH --nodes 1
#SBATCH --ntasks 1
#SBATCH --cpus-per-task 3
#SBATCH --mem 8G
#SBATCH --time 05:10:00
#SBATCH --output /users/vsomervi/scratch/vsomervi/logs/%x_%j.out
#SBATCH --error /users/vsomervi/scratch/vsomervi/logs/%x_%j.err

####--------------------------------------
##preparation
##set you bash variables in order to quickly call them in the script
####--------------------------------------
username=vsomervi


hmmsss=DruE
genomesss=Lapis_ESL0263
hmmsss=DruA
genomesss=Lbombicola_ESL0230
####---------------------
##modules
####--------------------------------------

module load HPC/Software 
module load SequenceAnalysis/HMM-Profile/hmmer/3.1b2

####--------------------------------------
##start of script
####--------------------------------------

start=$SECONDS
echo "Step: Test"

echo "=========================================================="
date +"START : %a %b %e %Y %H:%M:%S "
#echo -e "Hello " ${username}
rm /users/vsomervi/scratch/vsomervi/SAGE_II/05_Identification_search/all.scans
rm  /users/vsomervi/scratch/vsomervi/SAGE_II/05_Identification_search/all.scans_oneline
for genomesss in $(ls /users/vsomervi/scratch/common_files/SAGE_II/Genomes/| grep -v "outgroup")
do

rm -r /users/vsomervi/scratch/vsomervi/SAGE_II/05_Identification_search/${genomesss}/
mkdir -p /users/vsomervi/scratch/vsomervi/SAGE_II/05_Identification_search/${genomesss}/stdouts

echo "=================================================================="
echo ${genomesss}

for hmmsss in $(ls /users/vsomervi/scratch/vsomervi/SAGE_II/04_HMMs_final/  | grep ".txt$" |sed 's/.txt//g')
do

echo "-----------------------------------------------------------------"
echo ${hmmsss}

#hmmpress /users/vsomervi/scratch/vsomervi/SAGE_II/04_HMMs/${hmmsss}.txt
hmmsearch --cpu 3 -E 0.001 --tblout /users/vsomervi/scratch/vsomervi/SAGE_II/05_Identification_search/${genomesss}/${hmmsss}.scan.tab -o /users/vsomervi/scratch/vsomervi/SAGE_II/05_Identification_search/${genomesss}/stdouts/${hmmsss}.scan   /users/vsomervi/scratch/vsomervi/SAGE_II/04_HMMs_final/${hmmsss}.txt /users/vsomervi/scratch/common_files/SAGE_II/Genomes/${genomesss}/*.faa

grep -v "^#" /users/vsomervi/scratch/vsomervi/SAGE_II/05_Identification_search/${genomesss}/${hmmsss}.scan.tab  |awk -v genomzz="$genomesss" -v hmmzz="$hmmsss" '{OFS="\t"}{print genomzz,hmmzz,$3,$19}' >> /users/vsomervi/scratch/vsomervi/SAGE_II/05_Identification_search/all.scans
q
#grep -v "^#" /users/vsomervi/scratch/vsomervi/SAGE_II/05_Identification/${genomesss}/${hmmsss}.scan.tab  |awk '{OFS="\t"}{print $3}'| sed -z 's/\n/,/g;s/,$/\n/' |awk -v genomzz="$genomesss" -v hmmzz="$hmmsss" '{OFS="\t"}{print genomzz,hmmzz,$1}'  >> /users/vsomervi/scratch/vsomervi/SAGE_II/05_Identification/all.scans_oneline

done # hmmsss
done #genomessss
###--------------------------
##outgroup
#for the outgroups we do it seperately
##---------------------------
for genomesss in $(ls /users/vsomervi/scratch/common_files/SAGE_II/Genomes/outgroup/| grep -v "tar")
do

rm -r /users/vsomervi/scratch/vsomervi/SAGE_II/05_Identification_search/${genomesss}/
mkdir -p /users/vsomervi/scratch/vsomervi/SAGE_II/05_Identification_search/${genomesss}/stdouts

echo "=================================================================="
echo ${genomesss}

for hmmsss in $(ls /users/vsomervi/scratch/vsomervi/SAGE_II/04_HMMs_final/ | grep ".txt$" |sed 's/.txt//g')
do

echo "-----------------------------------------------------------------"
echo ${hmmsss}


#hmmpress /users/vsomervi/scratch/vsomervi/SAGE_II/04_HMMs/${hmmsss}.txt
hmmsearch --cpu 3 -E 0.001 --tblout /users/vsomervi/scratch/vsomervi/SAGE_II/05_Identification_search/${genomesss}/${hmmsss}.scan.tab -o /users/vsomervi/scratch/vsomervi/SAGE_II/05_Identification_search/${genomesss}/stdouts/${hmmsss}.scan   /users/vsomervi/scratch/vsomervi/SAGE_II/04_HMMs_final/${hmmsss}.txt /users/vsomervi/scratch/common_files/SAGE_II/Genomes/outgroup/${genomesss}/*.faa

grep -v "^#" /users/vsomervi/scratch/vsomervi/SAGE_II/05_Identification_search/${genomesss}/${hmmsss}.scan.tab  |awk -v genomzz="$genomesss" -v hmmzz="$hmmsss" '{OFS="\t"}{print genomzz,hmmzz,$3,$19}' >> /users/vsomervi/scratch/vsomervi/SAGE_II/05_Identification_search/all.scans


#grep -v "^#" /users/vsomervi/scratch/vsomervi/SAGE_II/05_Identification/${genomesss}/${hmmsss}.scan.tab  |awk '{OFS="\t"}{print $3}'| sed -z 's/\n/,/g;s/,$/\n/' |awk -v genomzz="$genomesss" -v hmmzz="$hmmsss" '{OFS="\t"}{print genomzz,hmmzz,$1}'  >> /users/vsomervi/scratch/vsomervi/SAGE_II/05_Identification/all.scans_oneline

done # hmmsss
done #genomessss

duration=$(( SECONDS - start ))
echo -e "The script ran for "${duration} "seconds"

##------------------------------
## genome names
##------------------------------
ls /users/vsomervi/scratch/common_files/SAGE_II/Genomes/ |grep -v "outgroup" > /users/vsomervi/scratch/vsomervi/SAGE_II/05_Identification_search/all.genomes
ls /users/vsomervi/scratch/common_files/SAGE_II/Genomes/outgroup/| grep -v "tar" >> /users/vsomervi/scratch/vsomervi/SAGE_II/05_Identification_search/all.genomes

```


Now have a look at the outputs

```{bash}
## complete output
less /users/vsomervi/scratch/vsomervi/SAGE_II/05_Identification_search/all.scans
#for which genomes did we find how many ortholgos
cut -f 1 /users/vsomervi/scratch/vsomervi/SAGE_II/05_Identification_search/all.scans |sort| uniq -c 
#how many orthogs were identified
cut -f 2 /users/vsomervi/scratch/vsomervi/SAGE_II/05_Identification_search/all.scans |sort| uniq -c 

# look at genome names
less /users/vsomervi/scratch/vsomervi/SAGE_II/05_Identification_search/all.genomes
```

Now transfer the output files locally for Data analysis. 

```{bash}
mkdir -p /home/vincent/Desktop/Projects/2021_SAGE_II/05_Identification_search/ 
scp  vsomervi@wally-front1.unil.ch:/users/vsomervi/scratch/vsomervi/SAGE_II/05_Identification_search/all.* /home/vincent/Desktop/Projects/2021_SAGE_II/05_Identification_search/ 


scp -r vsomervi@wally-front1.unil.ch:/users/vsomervi/scratch/vsomervi/SAGE_II/04_HMMs_final/ /home/vincent/Desktop/Projects/2021_SAGE_II/


```

Here, I merge all Hmms into a new folder to bring them locally,

```{bash}
rm -r  /users/vsomervi/scratch/vsomervi/SAGE_II/04_HMMs_final/
mkdir -p /users/vsomervi/scratch/vsomervi/SAGE_II/04_HMMs_final/
for hmmsss in $(ls /users/vsomervi/scratch/vsomervi/SAGE_II/04_HMMs/ | grep ".txt$" |sed 's/.txt//g')
do
cp /users/vsomervi/scratch/vsomervi/SAGE_II/04_HMMs/${hmmsss}.txt /users/vsomervi/scratch/vsomervi/SAGE_II/04_HMMs_final/${hmmsss}.txt

done
##----------------------------------
## gather alessandros (R-M)
#no genes for type 2 and type 3
#did not work...I used e.rochas HMMs
##----------------------------------

#for hmmsss in $(ls /users/vsomervi/scratch/abrandul/SAGE_II/HMM/ | grep ".txt$" |sed 's/.txt//g' |cut -d '-' -f 1)
#do

#cp /users/vsomervi/scratch/abrandul/SAGE_II/HMM//${hmmsss}-*.txt /users/vsomervi/scratch/vsomervi/SAGE_II/04_HMMs_final/${hmmsss}.txt

#done

##----------------------------------
## gather Lucas (Retrons)
##----------------------------------


for hmmsss in $(ls /users/vsomervi/scratch/lculebra/SAGE_II/HMM/ | grep ".txt$" |sed 's/.txt//g' |awk -F "_HMM" '{print $1}'|sed 's/_COG3344//g')
do

cp /users/vsomervi/scratch/lculebra/SAGE_II/HMM//${hmmsss}_*.txt /users/vsomervi/scratch/vsomervi/SAGE_II/04_HMMs_final/${hmmsss}.txt

done


##----------------------------------
## gather fabrice (Zorya)
##----------------------------------


for hmmsss in $(ls /users/vsomervi/scratch/fbattist/sage2/hmm/ | grep ".txt$" |sed 's/.txt//g' |cut -d '_' -f 1)
do

cp /users/vsomervi/scratch/fbattist/sage2/hmm//${hmmsss}_*.txt /users/vsomervi/scratch/vsomervi/SAGE_II/04_HMMs_final/${hmmsss}.txt

done

```




### 5. Analyse and Plot data

Here, we take the output and plot it. 

```{r eval=TRUE}
library(readr)
library(tidyverse)
all_scans <- read_delim("~/Desktop/Projects/2021_SAGE_II/05_Identification_search/all.scans", "\t", escape_double = FALSE, col_names = c("genome","geneOfInterest","gene")) %>% filter(!geneOfInterest %in% c("GalA","GalB"))
all_scans$geneOfInterest <- plyr::revalue(all_scans$geneOfInterest,c("JatA"="JetA"))

table(all_scans$geneOfInterest)
all_genomes <- read_delim("~/Desktop/Projects/2021_SAGE_II/05_Identification_search/all.genomes", "\t", escape_double = FALSE, col_names = c("genome"))

#check if all genomes are present

all(all_genomes$genome %in% all_scans$genome)
all_genomes[which(!all_genomes$genome %in% all_scans$genome),]

###count presence of gene per genome

all_scans_count <- all_scans %>% group_by(genome,geneOfInterest) %>% dplyr::summarise(counts=n())

```

Next we want to know which genes are necessary for which defense mechanisms and/or subtype. 
To do this we need to Download the .csv from the [google.sheets with all our entries](https://docs.google.com/spreadsheets/d/1Gp4cqbiYJ34fB7wbcqJBKsGxwgaMyPRrETnVK9PS_4M/edit#gid=0)

```{r eval=TRUE}
phagedefenseGenes <- read_csv("~/Desktop/Projects/2021_SAGE_II/cleaned_group9_phagedefense_cleaned.csv", skip = 1) %>% select( mechanism,subtype,gene)

##paste_subtpye and mechanism column 

phagedefenseGenes$mechanism_final <- paste(phagedefenseGenes$mechanism,phagedefenseGenes$subtype,sep="-") %>% gsub("-NA","",.)

phagedefenseGenes_count <- phagedefenseGenes %>% group_by(mechanism_final) %>% dplyr::summarise(mechanism_gene_count=n())

phagedefenseGenes_02 <- phagedefenseGenes %>% select(mechanism_final,gene)

phagedefenseGenes_cleaned <- merge(phagedefenseGenes_02,phagedefenseGenes_count)

```

Now we want to merge the information from the genomes with the information from the mechanism description.

```{r eval=TRUE}
see_gene_count <- merge(all_scans_count,phagedefenseGenes_cleaned,by.x="geneOfInterest",by.y="gene",all.x=TRUE) 

tmp <- see_gene_count %>% filter(is.na(mechanism_final))
# table(tmp$geneOfInterest)

pathway_identification <- see_gene_count %>% group_by(genome,mechanism_final,mechanism_gene_count) %>%  summarise(number_genes_identified=n())

pathway_identification$percentPathway <- 100*(pathway_identification$number_genes_identified/pathway_identification$mechanism_gene_count)

```


Now we want to plot this data

```{r eval=TRUE}

## Remember we have the following meachanisms that we are looking at. 
unique(phagedefenseGenes_02$mechanism_final)

#order the pathways according to these
pathway_identification$mechanism_final  = factor(pathway_identification$mechanism_final , levels=unique(phagedefenseGenes_02$mechanism_final))

##-------------------------------

ggheatmap <- ggplot(pathway_identification, aes(y=genome, x=mechanism_final, fill = percentPathway))+
 geom_tile(color = "white",size=1.1)+
 scale_fill_gradient2(low = "white", high = "red", mid = "white",
   midpoint = 0, limit = c(-100,100), space = "Lab",
   name="max growth") +
  theme_minimal()+ # minimal theme
  scale_x_discrete(position = "top") +
  # labs(x=expression(italic("S. thermophilus")),y=expression(italic("L. delbrueckii")))+
 theme(axis.text.y = element_text(size=7),
      axis.text.x = element_text(angle = 45,vjust=1,
    size = 12, hjust = 0),axis.title=element_blank(), 
    legend.title = element_blank())
 # coord_fixed()
# Print the heatmap
print(ggheatmap)
  

###-------------------------------------
#make presence/absence
###-------------------------------------

pathway_identification_filter <- pathway_identification%>% filter(!mechanism_final %in% c("Disarm-core","Zorya-core")) %>% mutate(Presence_final = ifelse(percentPathway!=100||is.na(percentPathway), 0, 1)) #%>%replace(is.na(Presence_final), 0) 

ggheatmap <- ggplot(pathway_identification_filter, aes(y=genome, x=mechanism_final, fill = Presence_final))+
 geom_tile(color = "white",size=1.1)+
 scale_fill_gradient2(na.value = 'white',low = "white", high = "red", mid = "white",
   midpoint = 0, limit = c(-1,1), space = "Lab",
   name="max growth") +
  theme_classic()+ # minimal theme
  scale_x_discrete(position = "top") +
  # labs(x=expression(italic("S. thermophilus")),y=expression(italic("L. delbrueckii")))+
 theme(axis.text.y = element_text(size=7),
      axis.text.x = element_text(angle = 45,vjust=1,
    size = 12, hjust = 0),axis.title=element_blank(), 
    legend.title = element_blank(),legend.position = "none")
 # coord_fixed()
# Print the heatmap
print(ggheatmap)

```


### 6. CRISPRcasTyping

First, we need to get the genomes locally. 

```{bash}
mkdir -p /home/vincent/Desktop/Projects/2021_SAGE_II/06_genomes/
scp  -r vsomervi@wally-front1.unil.ch:/users/vsomervi/scratch/common_files/SAGE_II/Genomes/ /home/vincent/Desktop/Projects/2021_SAGE_II/06_genomes/ 
```

Then we annotate/extract the CRISPR regions online with [CRISPRcasTyper](cctyper.crispr.dk) explained [here](https://www.liebertpub.com/doi/10.1089/crispr.2020.0059). 
You can upload the *.fna file within the sample folder and run CRISPRcasTyper with default settings. 
We can split up the search for the genomes and fill in the [google.sheets](https://docs.google.com/spreadsheets/d/1Gp4cqbiYJ34fB7wbcqJBKsGxwgaMyPRrETnVK9PS_4M/edit#gid=732249118).
from these annotations we can print the absence presence of CRISPR-cas systems in the genomes. 

##### Analysis

After everyone has filled in the CRISPR sheet we can download the .csv and plot the data. 
Try this out yourself. 

Some hints:

1. Download the google.sheet as csv. 
2. import it into r
3. analyse the Prediction column

```{r}

library(tidyverse)
group9_phagedefense_CRISPR_scan <- read_csv("~/Desktop/Projects/2021_SAGE_II/CRISPR_annotation/group9_phagedefense_CRISPR_scan.csv",skip = 1)


defense_long <- group9_phagedefense_CRISPR_scan %>% select(genome,Prediction)

group9_phagedefense_CRISPR_scan %>% filter(!is.na(Prediction))%>% group_by(genome,Prediction) %>% summarise(counts=n()) %>% select(-count) %>% spread(.,)
  
  
  
snps_freebayes_long <- gather(snps_freebayes, sample, Snps, colnames(snps_freebayes)[6]:colnames(snps_freebayes)[ncol(snps_freebayes)], factor_key=TRUE,na.rm = TRUE) 
sample_relative_safe_final_tsne <- spread(sample_relative_safe_final_prep, sample, allel_frequency)


ggheatmap <- ggplot(defense_long, aes(y=genome, x=Prediction, fill = percentPathway))+
 geom_tile(color = "white",size=1.1)+
 scale_fill_gradient2(low = "white", high = "red", mid = "white",
   midpoint = 0, limit = c(-100,100), space = "Lab",
   name="max growth") +
  theme_minimal()+ # minimal theme
  scale_x_discrete(position = "top") +
  # labs(x=expression(italic("S. thermophilus")),y=expression(italic("L. delbrueckii")))+
 theme(axis.text.y = element_text(size=7),
      axis.text.x = element_text(angle = 45,vjust=1,
    size = 12, hjust = 0),axis.title=element_blank(), 
    legend.title = element_blank())
 # coord_fixed()
# Print the heatmap
print(ggheatmap)
  


```


### Phylogeny

Here, I use the phylogeny to order the defense mechanisms

```{r}


###-----------------------------------------------
##preperation
###-----------------------------------------------

defense_long_complete_tmp <- pathway_identification%>% filter(!mechanism_final %in% c("Disarm-core","Zorya-core")) %>% mutate(Presence_final = ifelse(percentPathway!=100||is.na(percentPathway), 0, 1))%>% select(genome,mechanism_final,Presence_final)  %>% spread(., mechanism_final, Presence_final) %>%replace(is.na(.), 0) %>% gather(., mechanism_final,Presence_final,"Druantia-I":"CBASS",  factor_key=TRUE,na.rm = TRUE)#%>%replace(is.na(Presence_final), 0) 

# pathway_identification_species_plot_complete <- spread(pathway_identification_species_plot, mechanism_final, percent_species) %>%replace(is.na(.), 0) %>% gather(., mechanism_final,percent_species,"R-M-I":"CRISPR type III_A",  factor_key=TRUE,na.rm = TRUE)


defense_long_complete <- defense_long %>%  group_by(genome,Prediction) %>% summarise(counts=n()) %>% select(-counts) %>% rename("mechanism_final"="Prediction") %>% add_column(Presence_final=1) %>% spread(.,mechanism_final,Presence_final) %>%replace(is.na(.), 0) %>% gather(., mechanism_final,Presence_final,"I-B":"Unknown",  factor_key=TRUE,na.rm = TRUE) %>% mutate(mechanism_final=paste0("CRISPR ",mechanism_final)) %>% filter(mechanism_final!="no CRISPR")
  
pathway_identification_filter <- rbind(defense_long_complete_tmp,defense_long_complete)



write.table(pathway_identification_filter,"~/Desktop/Projects/2021_SAGE_II/Phylogeny/defense_mechanism_presence.txt",sep="\t",quote = FALSE,row.names = FALSE)
###-----------------------------------------------
##plotting
###-----------------------------------------------
library(ape)
library(ggtree)
library(tidyverse)
library(tidytree)
library(treeio)
library(TreeTools)

Bee_broad_tree_raw <- read.nexus("~/Desktop/Projects/2021_SAGE_II/Phylogeny/Stingless_BEE_microbiome_Rooted.tree._rerooted.nwk", tree.names = NULL, force.multi = FALSE)


gtree_01 <- ggtree(Bee_broad_tree_raw) +
  geom_tiplab(align = TRUE,size=3) + # tiplabls aligned
  geom_rootedge(rootedge = 0)  #NO ROOT  
  gtree_01


  ###--------------------------------
#add all
###--------------------------------
  
  pathway_identification_filter <- read_delim("~/Desktop/Projects/2021_SAGE_II/Phylogeny/defense_mechanism_presence.txt", 
    "\t", escape_double = FALSE, trim_ws = TRUE)
  
 pathway_identification_filter_sub  <- pathway_identification_filter %>% select(genome,mechanism_final,Presence_final) %>% rename("label"="genome")
  
as_tibble(Bee_broad_tree_raw)$label %in% pathway_identification_filter_sub$label
pathway_identification_filter_sub$label %in% as_tibble(Bee_broad_tree_raw)$label

pathway_identification_filter_sub[which(!pathway_identification_filter_sub$label %in% as_tibble(Bee_broad_tree_raw)$label),"label"] %>% unique()
 

ptotal <- facet_plot(gtree_01, panel = 'R-M', data = pathway_identification_filter_sub, 
                geom = geom_tile,color = "grey",width=0.8,height=0.8,size=0.2,
                mapping = aes(x = mechanism_final,fill=Presence_final))+scale_x_discrete()+scale_fill_gradient2(na.value="white",low = "white", high = "red", mid = "white",
   midpoint = 0, limit = c(-1,1), space = "Lab",
   name="max growth")+theme_classic()+theme(axis.text.y = element_text(size=7),
      axis.text.x = element_text(angle = 45,size=10),axis.title=element_blank(), 
    legend.title = element_blank(),legend.position = "none",strip.background = element_blank(),
  strip.text.x = element_blank()) #+xlim_expand(c(0,500), 'Genome count')
ptotal
  
  ###--------------------------------
#add R-M
colorss="red"
###--------------------------------
table(pathway_identification_filter$mechanism_final)
 pathway_identification_filter_sub  <- pathway_identification_filter %>% select(genome,mechanism_final,Presence_final) %>% rename("label"="genome") %>% filter(mechanism_final %in% c("R-M-IIG","R-M-III","R-M-II","R-M-I","R-M-IV","BREX","Disarm-I","Disarm-II")) %>% mutate(type = ifelse(Presence_final==1, "R-M", NA)) # %>% filter(Presence_final==1)
  
as_tibble(Bee_broad_tree_raw)$label %in% pathway_identification_filter_sub$label
pathway_identification_filter_sub$label %in% as_tibble(Bee_broad_tree_raw)$label

pathway_identification_filter_sub[which(!pathway_identification_filter_sub$label %in% as_tibble(Bee_broad_tree_raw)$label),"label"] %>% unique()
 

pathway_identification_filter_sub$mechanism_final  = factor(pathway_identification_filter_sub$mechanism_final , levels=c(
  "R-M-I", "R-M-II", "R-M-IIG", "R-M-III", "R-M-IV",
  "BREX","Disarm-I","Disarm-II", 
  "Druantia-I",  "Druantia-II","Druantia-III","Hachiman","Kiwa","Lamassu","Shedu","Thoeris","Zorya-I","Zorya-II","Wadjet-I","Septu",
  "CBASS","Abi",
  "GTPase","dNTPaminase","Retrons",
  
  
  
))

# p2 <- facet_plot(gtree_01, panel = 'R-M', data = pathway_identification_filter_sub, 
#                 geom = geom_tile,color = "grey",width=0.8,height=0.8,size=1.0,
#                 mapping = aes(x = mechanism_final,fill=Presence_final))+scale_x_discrete()+scale_fill_gradient2(na.value="white",low = "white", high = colorss, mid = "white",
#    midpoint = 0, limit = c(-1,1), space = "Lab",
#    name="max growth")+theme_classic()+theme(axis.text.y = element_text(size=7),
#       axis.text.x = element_text(angle = 45,size=10),axis.title=element_blank(), 
#     legend.title = element_blank(),legend.position = "none",strip.background = element_blank(),
#   strip.text.x = element_blank()) #+xlim_expand(c(0,500), 'Genome count')
# p2


p2 <- facet_plot(gtree_01, panel = 'R-M', data = pathway_identification_filter_sub, 
                geom = geom_tile,color = "grey",width=0.8,height=0.8,size=0.2,
                mapping = aes(x = mechanism_final,fill=type))+scale_fill_manual(values=colorss)+scale_x_discrete()+theme_classic()+theme(axis.text.y = element_text(size=7),
      axis.text.x = element_text(angle = 45,size=10,hjust=1),axis.title=element_blank(), 
    legend.title = element_blank(),legend.position = "none",strip.background = element_blank(),
  strip.text.x = element_blank()) #+xlim_expand(c(0,500), 'Genome count')
p2



  ###--------------------------------
#add unkown mechanism
colorss=c("blue","red")
###--------------------------------
table(pathway_identification_filter$mechanism_final)
 pathway_identification_filter_sub  <- pathway_identification_filter %>% select(genome,mechanism_final,Presence_final) %>% rename("label"="genome") %>% filter(mechanism_final %in% c("Druantia-I",  "Druantia-II","Druantia-III","Hachiman","Kiwa","Lamassu","Shedu","Thoeris","Zorya-I","Zorya-II","Wadjet-I","Septu")) %>% mutate(type = ifelse(Presence_final==1, "Unknown", NA)) #%>% add_column(type="Unknown") %>% filter(Presence_final==1)
  
as_tibble(Bee_broad_tree_raw)$label %in% pathway_identification_filter_sub$label
pathway_identification_filter_sub$label %in% as_tibble(Bee_broad_tree_raw)$label

pathway_identification_filter_sub[which(!pathway_identification_filter_sub$label %in% as_tibble(Bee_broad_tree_raw)$label),"label"] %>% unique()
 

pathway_identification_filter_sub$mechanism_final  = factor(pathway_identification_filter_sub$mechanism_final , levels=c(
  "R-M-I", "R-M-II", "R-M-IIG", "R-M-III", "R-M-IV",
  "BREX","Disarm-I","Disarm-II", 
  "Druantia-I",  "Druantia-II","Druantia-III","Hachiman","Kiwa","Lamassu","Shedu","Thoeris","Zorya-I","Zorya-II","Wadjet-I","Septu",
  "CBASS","Abi",
  "GTPase","dNTPaminase","Retrons"
  
  
))

# p3 <- facet_plot(p2, panel = 'R-M', data = pathway_identification_filter_sub, 
#                 geom = geom_tile,color = "grey",width=0.8,height=0.8,size=1.0,
#                 mapping = aes(x = mechanism_final,fill=Presence_final))+scale_x_discrete()+scale_fill_gradient2(na.value="white",low = "white", high = colorss, mid = "white",
#    midpoint = 0, limit = c(-1,1), space = "Lab",
#    name="max growth")+theme_classic()+theme(axis.text.y = element_text(size=7),
#       axis.text.x = element_text(angle = 45,size=10),axis.title=element_blank(), 
#     legend.title = element_blank(),legend.position = "none",strip.background = element_blank(),
#   strip.text.x = element_blank()) #+xlim_expand(c(0,500), 'Genome count')
# p3


p3 <- facet_plot(p2, panel = 'unknown', data = pathway_identification_filter_sub, 
                geom = geom_tile,color = "grey",width=0.8,height=0.8,size=0.2,
                mapping = aes(x = mechanism_final,fill=type))+scale_fill_manual(values=colorss)+scale_x_discrete()+theme_classic()+theme(axis.text.y = element_text(size=7),
      axis.text.x = element_text(angle = 45,size=10,hjust=1),axis.title=element_blank(), 
    legend.title = element_blank(),legend.position = "none",strip.background = element_blank(),
  strip.text.x = element_blank()) #+xlim_expand(c(0,500), 'Genome count')
p3

 ###--------------------------------
#ABI
colorss=c("blue","red","green")
###--------------------------------
table(pathway_identification_filter$mechanism_final)
 pathway_identification_filter_sub  <- pathway_identification_filter %>% select(genome,mechanism_final,Presence_final) %>% rename("label"="genome") %>% filter(mechanism_final %in% c("CBASS","Abi")) %>% mutate(type = ifelse(Presence_final==1, "Abi", NA)) #%>% add_column(type="Unknown") %>% filter(Presence_final==1)
  
as_tibble(Bee_broad_tree_raw)$label %in% pathway_identification_filter_sub$label
pathway_identification_filter_sub$label %in% as_tibble(Bee_broad_tree_raw)$label

pathway_identification_filter_sub[which(!pathway_identification_filter_sub$label %in% as_tibble(Bee_broad_tree_raw)$label),"label"] %>% unique()
 

pathway_identification_filter_sub$mechanism_final  = factor(pathway_identification_filter_sub$mechanism_final , levels=c(
  "R-M-I", "R-M-II", "R-M-IIG", "R-M-III", "R-M-IV",
  "BREX","Disarm-I","Disarm-II", 
  "Druantia-I",  "Druantia-II","Druantia-III","Hachiman","Kiwa","Lamassu","Shedu","Thoeris","Zorya-I","Zorya-II","Wadjet-I","Septu",
  "CBASS","Abi",
  "GTPase","dNTPaminase","Retrons"
  
  
))

p4 <- facet_plot(p3, panel = 'Abi', data = pathway_identification_filter_sub, 
                geom = geom_tile,color = "grey",width=0.8,height=0.8,size=0.2,
                mapping = aes(x = mechanism_final,fill=type))+scale_fill_manual(values=colorss)+scale_x_discrete()+theme_classic()+theme(axis.text.y = element_text(size=7),
      axis.text.x = element_text(angle = 45,size=10,hjust=1),axis.title=element_blank(), 
    legend.title = element_blank(),legend.position = "none",strip.background = element_blank(),
  strip.text.x = element_blank()) #+xlim_expand(c(0,500), 'Genome count')
p4

 ###--------------------------------
#Novel
colorss=c("blue","red","green","salmon")
###--------------------------------
table(pathway_identification_filter$mechanism_final)
 pathway_identification_filter_sub  <- pathway_identification_filter %>% select(genome,mechanism_final,Presence_final) %>% rename("label"="genome") %>% filter(mechanism_final %in% c("GTPase","dNTPaminase","Retrons")) %>% mutate(type = ifelse(Presence_final==1, "novel", NA)) #%>% add_column(type="Unknown") %>% filter(Presence_final==1)
  
as_tibble(Bee_broad_tree_raw)$label %in% pathway_identification_filter_sub$label
pathway_identification_filter_sub$label %in% as_tibble(Bee_broad_tree_raw)$label

pathway_identification_filter_sub[which(!pathway_identification_filter_sub$label %in% as_tibble(Bee_broad_tree_raw)$label),"label"] %>% unique()
 

pathway_identification_filter_sub$mechanism_final  = factor(pathway_identification_filter_sub$mechanism_final , levels=c(
  "R-M-I", "R-M-II", "R-M-IIG", "R-M-III", "R-M-IV",
  "BREX","Disarm-I","Disarm-II", 
  "Druantia-I",  "Druantia-II","Druantia-III","Hachiman","Kiwa","Lamassu","Shedu","Thoeris","Zorya-I","Zorya-II","Wadjet-I","Septu",
  "CBASS","Abi",
  "GTPase","dNTPaminase","Retrons"
  
  
))

# library(ggplot2)
library(ggh4x)
p5 <- facet_plot(p4, panel = 'Novel', data = pathway_identification_filter_sub, 
                geom = geom_tile,color = "grey",width=0.8,height=0.8,size=0.2,
                mapping = aes(x = mechanism_final,fill=type))+scale_fill_manual(values=colorss)+scale_x_discrete()+theme_classic()+theme(axis.text.y = element_text(size=7),
      axis.text.x = element_text(angle = 45,size=10,hjust=1),axis.title=element_blank(), 
    legend.title = element_blank(),legend.position = "none",strip.background = element_blank(),
  strip.text.x = element_blank()) +force_panelsizes(cols = c(1,0.8,1.1,0.2,0.3))#+xlim_expand(c(0,500), 'Genome count')
p5

###--------------------------------
#CRISPR
colorss=c("blue","red","green","salmon","gold")
###--------------------------------
table(pathway_identification_filter$mechanism_final)
 pathway_identification_filter_sub  <- pathway_identification_filter %>% select(genome,mechanism_final,Presence_final) %>% rename("label"="genome") %>% filter(mechanism_final %in% c("CRISPR I-B",     "CRISPR I-C",     "CRISPR I-E",    "CRISPR II-A",   "CRISPR III-A", "CRISPR Unknown")) %>% mutate(type = ifelse(Presence_final==1, "CRISPR", NA)) #%>% add_column(type="Unknown") %>% filter(Presence_final==1)
  
as_tibble(Bee_broad_tree_raw)$label %in% pathway_identification_filter_sub$label
pathway_identification_filter_sub$label %in% as_tibble(Bee_broad_tree_raw)$label

pathway_identification_filter_sub[which(!pathway_identification_filter_sub$label %in% as_tibble(Bee_broad_tree_raw)$label),"label"] %>% unique()
 

pathway_identification_filter_sub$mechanism_final  = factor(pathway_identification_filter_sub$mechanism_final , levels=c(
  "R-M-I", "R-M-II", "R-M-IIG", "R-M-III", "R-M-IV",
  "BREX","Disarm-I","Disarm-II", 
  "Druantia-I",  "Druantia-II","Druantia-III","Hachiman","Kiwa","Lamassu","Shedu","Thoeris","Zorya-I","Zorya-II","Wadjet-I","Septu",
  "CBASS","Abi",
  "GTPase","dNTPaminase","Retrons",
  "CRISPR I-B",     "CRISPR I-C",     "CRISPR I-E",    "CRISPR II-A",   "CRISPR III-A", "CRISPR Unknown"
  
  
))

# library(ggplot2)
library(ggh4x)
p6 <- facet_plot(p5, panel = 'CRISPR', data = pathway_identification_filter_sub, 
                geom = geom_tile,color = "grey",width=0.8,height=0.8,size=0.2,
                mapping = aes(x = mechanism_final,fill=type))+scale_fill_manual(values=colorss)+scale_x_discrete()+theme_classic()+theme(axis.text.y = element_text(size=7),
      axis.text.x = element_text(angle = 45,size=10,hjust=1),axis.title=element_blank(), 
    legend.title = element_blank(),legend.position = "none",strip.background = element_blank(),
  strip.text.x = element_blank()) +force_panelsizes(cols = c(2,0.8,1.1,0.2,0.3,0.6))
p6


  png("~/Desktop/Projects/2021_SAGE_II/heatmap_with_phylogeny.png", width = 3500, height = 2000,res=300)


p6
 dev.off()
```

